sys/kern/vm_map.c

```diff
@@ -161,6 +161,46 @@ void vm_segment_destroy(vm_map_t *map, vm_segment_t *seg) {
   vm_segment_free(seg);
 }
 
+/* there are few cases:
+ * 1) we want to free the whole segment
+ * 2) we want to free the beginning of the segment and leave the end
+ * without changes
+ * 3) we want to free the end of the segment and leave the beginning
+ * without changes
+ * 4) we want to free the middle of the segment, so we want to split it to
+ * the new two segments with the hole between them */
+void vm_segment_destroy_range(vm_map_t *map, vm_segment_t *seg, vaddr_t start,
+                              vaddr_t end) {
+  assert(mtx_owned(&map->mtx));
+
+  /* case 1) */
+  if (seg->start == start && seg->end == end) {
+    vm_segment_destroy(map, seg);
+  } else if (seg->start == start) { /* case 2) */
+    assert(end < seg->end);
+    size_t length = end - seg->start;
+    vm_object_remove_range(seg->object, 0, length);
+    /* TODO there's no reference to pmap in page, so we have to do it here */
+    pmap_remove(map->pmap, seg->start, end);
+    seg->start = end;
```

Comment:
It is not clear for me, why we can not do it in the same way as in case 4?

Reply:
We cannot do it in the same way as in case 4 because in case we destroy the middle of the segment and finally we split it to the separate two segments. Whereas in case 2 we just destroy the beginning of the segment and change the start value to the new one.